### ISSUE TITLE
How to visualize input Dataloader after Augmentation?

### ISSUE THREAD

**Issue by chokyungjin** [ultralytics/yolov5#779]
Hi! I have a question your Dataloader.
Your yolov5 is soo difficult to me to analyze.. 
How to visualize input Dataloader after Augmentation? 
Before I training, I wonder how images come out in the dataloader.

Thanks!

**Comment by Ownmarc** [comment]
Trust me, Glenn’s code is really easy to read and understand compared to many repo in ML.

You can check the images created in the folder ‘’runs/exp0/‘’ they will show you 3 first batches when starting a training. Else you could simply create a dataloader object in a notebook or something using the repo’s code and load image one after an other and view them however you want by writing to disk or showing them with pyplot

**Comment by DacDinh147** [comment]
You have to look in the `LoadImagesAndLabels(Dataset)` class. This class return a tuple `(torch.from_numpy(img), labels_out, self.img_files[index], shapes)`.

Here `torch.from_numpy(img)` is a tensor with shape `[batch size, channels, w, h]` and `labels_out` is a list of item. Each item of `labels_out` represents a bounding box annotation for each image and is a list of bounding box `(x_center, y_center, w, h)` corresponding to the object in the specific image. Because the bounding box coordinates `(x_center, y_center, w, h)` was **normalized**, so you have to multiply it with the width and height of the image to visualize it `xc, yc, w, h = xc*w_img, yc*h_img, w*w_img, h*h_img `

Visualization with `batch size=6`

![image](https://drive.google.com/uc?export=view&id=1wgdTna707ojewRStTsFajv8HfRCZYsia)

**Comment by glenn-jocher** [comment]
We display this clearly in the notebook, suggest you start from there:
https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb#scrollTo=DLI1JmHU7B0l

![image](https://user-images.githubusercontent.com/26833433/90604617-fb4fe680-e1b1-11ea-8cbb-ee6a67978a9f.png)

**Comment by buimanhlinh96** [comment]
Hi @glenn-jocher, how can i add albumentations in torch library for augmentations in DataLoader?

**Comment by glenn-jocher** [comment]
@buimanhlinh96 you can customize the trainloader here:
https://github.com/ultralytics/yolov5/blob/7220cee1d1dc1f14003dbf8d633bbb76c547000c/utils/datasets.py#L328

**Comment by Auth0rM0rgan** [comment]
@glenn-jocher, Could you please provide more details that how to use [albumentations](https://github.com/albumentations-team/albumentations) in the DataLoader??

**Comment by glenn-jocher** [comment]
@Auth0rM0rgan we don't have an integration with albumentations. YOLOv5 augmentation hyperparameters are set here:
https://github.com/ultralytics/yolov5/blob/f6b3c96966d963a4c69736c8691f068e5b554247/data/hyp.scratch.yaml#L22-L33


You are free to modify the dataloader also as you see fit here:
https://github.com/ultralytics/yolov5/blob/7220cee1d1dc1f14003dbf8d633bbb76c547000c/utils/datasets.py#L328

**Comment by Auth0rM0rgan** [comment]
Hey @glenn-jocher, Thanks for the quick reply!

I know about the YOLOv5 augmentation but I'm trying to add some blur augmentations that are not in the yolov5 augmentation such as MotionBlur, Gaussian Blur, and ... but I have difficulty to add these augmentations inside YOLOv5. I will appreciate if you can help me even a little...

Gracias!

**Comment by glenn-jocher** [comment]
@Auth0rM0rgan ah I see. Well you can access the raw image as a numpy array `img` here after it's already passed through the YOLOv5 augmentations, so I would just insert any augmentations here. Be careful to correctly modify `labels` correspondingly as well if your augmentations require it though.
https://github.com/ultralytics/yolov5/blob/f6b3c96966d963a4c69736c8691f068e5b554247/utils/datasets.py#L526-L528

**Comment by Auth0rM0rgan** [comment]
@glenn-jocher Thanks! made it work. Now I can apply a set of different kinds of augmentations at pixel-level and spatial-level!

**Comment by glenn-jocher** [comment]
@Auth0rM0rgan hey cool! Maybe you could consider submitting a PR for general Albumentations integration? I know its a popular tool so others might find that useful.

**Comment by Auth0rM0rgan** [comment]
@glenn-jocher, Sure! will do it during the weekend!

**Comment by glenn-jocher** [comment]
@Auth0rM0rgan @buimanhlinh96 see PR #3882 for a proposed automatic Albumentations integration.

**Comment by glenn-jocher** [comment]
@Auth0rM0rgan @buimanhlinh96 @Ownmarc good news 😃! Your original issue may now be fixed ✅ in PR #3882. This PR implements a [YOLOv5](https://github.com/ultralytics/yolov5) 🚀 + [Albumentations](https://github.com/albumentations-team/albumentations) integration. The integration will automatically apply Albumentations transforms during YOLOv5 training if `albumentations>=1.0.0` is installed in your environment.

## Get Started

To use albumentations simply `pip install -U albumentations` and then update the augmentation pipeline as you see fit in the `Albumentations` class in `yolov5/utils/augmentations.py`. Note these Albumentations operations run **in addition to** the YOLOv5 **hyperparameter** augmentations, i.e. defined in [hyp.scratch.yaml](https://github.com/ultralytics/yolov5/blob/master/data/hyps/hyp.scratch.yaml).

```python
class Albumentations:
    # YOLOv5 Albumentations class (optional, used if package is installed)
    def __init__(self):
        self.transform = None
        try:
            import albumentations as A
            check_version(A.__version__, '1.0.0')  # version requirement

            self.transform = A.Compose([
                A.Blur(p=0.1),
                A.MedianBlur(p=0.1),
                A.ToGray(p=0.01)],
                bbox_params=A.BboxParams(format='yolo', label_fields=['class_labels']))

            logging.info(colorstr('albumentations: ') + ', '.join(f'{x}' for x in self.transform.transforms))
        except ImportError:  # package not installed, skip
            pass
        except Exception as e:
            logging.info(colorstr('albumentations: ') + f'{e}')

    def __call__(self, im, labels, p=1.0):
        if self.transform and random.random() < p:
            new = self.transform(image=im, bboxes=labels[:, 1:], class_labels=labels[:, 0])  # transformed
            im, labels = new['image'], np.array([[c, *b] for c, b in zip(new['class_labels'], new['bboxes'])])
        return im, labels
```

## Example Result

Example `train_batch0.jpg` on COCO128 dataset with Blur, MedianBlur and ToGray. See the YOLOv5 [Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb) to reproduce: <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>

![train_batch0](https://user-images.githubusercontent.com/26833433/124400879-ff331b80-dd25-11eb-9b67-fe85ac4ca104.jpg)


# Update

To receive this YOLOv5 update:

- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload with `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!
